packages/mds-compliance-api/handlers/get-violation-periods.ts

```diff
@@ -0,0 +1,144 @@
+import { ComplianceServiceClient, ComplianceSnapshotDomainModel } from '@mds-core/mds-compliance-service'
+import db from '@mds-core/mds-db'
+import { providerName, providers } from '@mds-core/mds-providers'
+import { ApiRequestQuery } from '@mds-core/mds-api-server'
+import express from 'express'
+import { parseRequest } from '@mds-core/mds-api-helpers'
+import { Policy, Timestamp } from '@mds-core/mds-types'
+import { isDefined, now, uuid } from '@mds-core/mds-utils'
+import { ComplianceAggregate, ComplianceApiRequest, ComplianceApiResponse, ComplianceViolationPeriod } from '../@types'
+
+export type ComplianceApiGetViolationPeriodsRequest = ComplianceApiRequest &
+  ApiRequestQuery<'start_time' | 'end_time' | 'provider_ids' | 'policy_ids'>
+
+export type ComplianceApiGetViolationPeriodsResponse = ComplianceApiResponse<{
+  start_time: Timestamp
+  end_time: Timestamp
+  results: ComplianceAggregate[]
+}>
+
+interface ComplianceAggregateMap {
+  [k: string]: {
+    complianceSnapshotViolationGroupings: ComplianceSnapshotDomainModel[][]
+    complianceSnapshots: null | ComplianceSnapshotDomainModel[]
+  }
+}
+
+async function convertComplianceSnapshotsArrayToComplianceViolationPeriod(
+  snapshots: ComplianceSnapshotDomainModel[]
+): Promise<ComplianceViolationPeriod> {
+  const compliance_array_response_id = uuid()
+  await ComplianceServiceClient.createComplianceArrayResponse({
+    compliance_array_response_id,
+    compliance_snapshot_ids: snapshots.map(s => s.compliance_snapshot_id),
+    provider_id: snapshots[0].provider_id
+  })
+  return {
+    start_time: snapshots[0].compliance_as_of,
+    end_time: snapshots[snapshots.length - 1].compliance_as_of,
+    snapshots_uri: `/compliance_snapshot_ids?token=${compliance_array_response_id}`
+  }
+}
+
+export const GetViolationPeriodsHandler = async (
+  req: ComplianceApiGetViolationPeriodsRequest,
+  res: ComplianceApiGetViolationPeriodsResponse,
+  next: express.NextFunction
+) => {
+  try {
+    const { scopes } = res.locals
+    const { start_time, end_time = now() } = parseRequest(req)
+      .single({ parser: Number })
+      .query('start_time', 'end_time')
+    if (!isDefined(start_time)) {
+      return res.status(400).send({ error: 'Missing required query param start_time' })
+    }
+    const { provider_ids, policy_ids } = parseRequest(req)
+      .single({ parser: (str: string | undefined | null) => str?.split(',') })
+      .query('provider_ids', 'policy_ids')
+
+    let p_provider_ids
+    if (scopes.includes('compliance:read:provider')) {
+      if (res.locals.claims && res.locals.claims.provider_id) {
+        const { provider_id } = res.locals.claims
+        p_provider_ids = [provider_id]
+      } else {
+        return res.status(403).send({ error: 'compliance:read:provider token missing provider_id' })
+      }
+    } else if (scopes.includes('compliance:read')) {
+      p_provider_ids = provider_ids ?? Object.keys(providers)
+    }
```

Comment:
Can you make this a closure so you don't need to use a `let`?

Reply:
Also probably wanna use camelCase instead of snake_case for new variables that aren't the exact names of entity properties

Reply:
I don't think I can make it a closure, but certainly there's a way to use a `const` instead